MSBLOC.Web/Controllers/AccountController.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using MongoDB.Driver;
+using MSBLOC.Core.Interfaces;
+using MSBLOC.Core.Services;
+using MSBLOC.Web.Interfaces;
+using MSBLOC.Web.Models;
+using Octokit;
+using IGitHubClientFactory = MSBLOC.Web.Interfaces.IGitHubClientFactory;
+
+namespace MSBLOC.Web.Controllers
+{
+    [Authorize]
+    public class AccountController : Controller
+    {
+        [HttpGet("~/signin")]
+        [AllowAnonymous]
+        public IActionResult SignIn()
+        {
+            return Challenge(new AuthenticationProperties { RedirectUri = "/" }, "GitHub");
+        }
+
+        [HttpGet("~/signout"), HttpPost("~/signout")]
+        public async Task<IActionResult> SignOut()
+        {
+            var authProperties = new AuthenticationProperties {RedirectUri = "/"};
+            await HttpContext.SignOutAsync(authProperties);
+            return SignOut(authProperties);
+        }
+
+        public async Task<IActionResult> ListRepositories([FromServices] IGitHubRepositoryContext repoContext, [FromServices] IGitHubClientFactory gitHubClientFactory)
+        {
+            var github = await gitHubClientFactory.CreateClientForCurrentUser();
+
+            var repositories = (await github.Repository.GetAllForCurrent())
+                .Select(r => new GitHubRepository
+                {
+                    Name = r.Name,
+                    Id = r.Id,
+                    Url = r.Url
+                })
+                .ToList();
+
+            var filter = Builders<GitHubRepository>.Filter.In(nameof(GitHubRepository.Id), repositories.Select(r => r.Id));
+
+            var savedRepos = await repoContext.Repositories.Find(filter).ToListAsync();
+
+            foreach (var savedRepo in savedRepos)
+            {
+                var repo = repositories.FirstOrDefault(r => r.Id == savedRepo.Id);
+                if (repo == null) continue;
+                repo.Id = savedRepo.Id;
+                repo.Secret = savedRepo.Secret;
+            }
+
+            return View(repositories);
+        }
+
+        [HttpGet]
```

Comment:
There should be a powershell function for this.

Reply:
Don't think so, the secret needs to be created by a logged in (to github) user. If a build system creates the secret we're not really authing anything.  That said, i think I'm going to change this over to use signed JWTs.

Reply:
Ok fair enough. Unless there is a GitHub api powershell module with a simple happy path to throw up a GitHub login the way the azure powershell cmdlets can through up a login popup modal.